org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -0,0 +1,495 @@
+        this(sslSocket, null, parentSocket, readChannel, writeChannel, engine);
+
+        // When there is no parent channel, this channel must be in
+        // blocking mode.
+        configureBlocking(true);
+    }
+
+    /**
+     * Give data already consumed by a call to the underlying socket's read
+     * method to this Socket, allowing it to be read by the SSLEngine.
+     */
+    public void setConsumedData(InputStream consumed) throws IOException {
+        if (consumed != null && consumed.available() > 0) {
+            this.consumed = consumed;
+            consumedChannel = Channels.newChannel(consumed);
+        }
+    }
+
+    /**
+     * Set whether or not to close the underlying Socket when the SSLSocket
+     * or this channel is closed.
+     */
+    public void setAutoClose(boolean on) {
+        autoClose = on;
+    }
+
+    /**
+     * Internal helper to bound the size of a blocking read to the maximum
+     * data available.
+     */
+    private int boundRead(int suggested) throws IOException {
+        return (int) boundRead((long) suggested);
+    }
+
+    /**
+     * Internal helper to bound the size of a blocking read to the maximum
+     * data available.
+     */
+    private long boundRead(long suggested) throws IOException {
+        // When there's consumed data left to read, ensure we bound by the
+        // amount available there before continuing.
+        if (consumed != null && consumed.available() > 0) {
+            return Math.min(consumed.available(), suggested);
+        }
+
+        // By setting consumed = null when consumed no longer has bytes
+        // available, we provide an easy check for which channel to read
+        // from.
+        consumed = null;
+        consumedChannel = null;
+
+        // If its a non-blocking underlying socket, then return suggested;
+        // it'll read as much as currently available.
+        if (!isBlocking()) {
+            return suggested;
+        }
+
+        // In both remaining cases (no channel or channel is blocking), bound
+        // the read above by the available data in the socket's input stream.
+        int available = parentSocket.getInputStream().available();
+        return Math.min(suggested, available);
+    }
+
+    public boolean finishConnect() throws IOException {
+        if (parent != null) {
+            if (!parent.finishConnect()) {
+                return false;
+            }
+        }
+
+        SSLEngineResult.HandshakeStatus state = engine.getHandshakeStatus();
+        if (state == SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING) {
+            return true;
+        }
+
+        int handshakeAttempts = 0;
+        int maxHandshakeAttempts = 100;
+
+        if (!isBlocking()) {
+            // When we're a non-blocking socket/channel, we'd farther rather
```

Comment:
Is this phrase correct?

Reply:
farther rather :laughing: Nah, it isn't. I'll update this.